src/compiler/checker.ts

```diff
@@ -981,6 +982,29 @@ namespace ts {
             }
         }
 
+        function checkAndReportErrorForUsingTypeAsValue(errorLocation: Node, name: string, meaning: SymbolFlags, nameNotFoundMessage: DiagnosticMessage, nameArg: string | Identifier): boolean {
+            const strictlyValueMeanings = SymbolFlags.Value & ~SymbolFlags.Type;
+            const strictlyTypeMeanings = SymbolFlags.Type & ~SymbolFlags.Value;
+
+            if (!(meaning & strictlyValueMeanings) || meaning & SymbolFlags.NamespaceModule) {
+                return false;
+            }
+
+            const nameAsType = resolveName(errorLocation, name, strictlyTypeMeanings, nameNotFoundMessage, nameArg);
+            if (!nameAsType) {
+                return false;
+            }
+
+            if (nameAsType.flags & SymbolFlags.Alias) {

```

Comment:
This seemed like it should have worked, but a symbol that resolves to a `NamespaceModule` meaning can also resolve to a bunch of type-inclusive meanings that will make `resolveAlias` happily accept the symbol. For example, in the test case `aliasOnMergedModuleInterface`, after subtracting the `NamespaceModule` meaning from my resolveName call, the symbol `foo` also has the `Interface` meaning, which means it satisfies `target.flags & meaning` in `resolveAlias` and returns the symbol. This, in turn, ends up triggering the new error message for a namespace symbol.